lib/ex_unit/lib/ex_unit/capture_server.ex

```diff
@@ -30,31 +34,29 @@ defmodule ExUnit.CaptureServer do
 
   def init(:ok) do
     state = %{
-      devices: {%{}, %{}},
+      devices: %{},
       log_captures: %{},
       log_status: nil
     }
 
     {:ok, state}
   end
 
-  def handle_call({:device_capture_on, name, pid}, _from, config) do
-    {names, refs} = config.devices
+  def handle_call({:device_capture_on, name, encoding, input}, {caller, _}, config) do
+    capture_device(name, encoding, input, config, caller)
+  end
 
-    if Map.has_key?(names, name) do
-      {:reply, {:error, :already_captured}, config}
-    else
-      orig_pid = Process.whereis(name)
-      Process.unregister(name)
-      Process.register(pid, name)
-      ref = Process.monitor(pid)
-      refs = Map.put(refs, ref, {name, orig_pid})
-      names = Map.put(names, name, true)
-      {:reply, {:ok, ref}, %{config | devices: {names, refs}}}
-    end
+  def handle_call({:device_output, name, ref}, _from, config) do
+    device = config.devices[name]
+    {_, output} = StringIO.contents(device.pid)
+    total = byte_size(output)
+    offset = device.refs[ref]
```

Comment:
Same as above, better error message (and I think better performance) in case something goes wrong:

```suggestion
    offset = Map.fetch!(device.refs, ref)
```

Reply:
I am honestly not super worried about this. It should never fail. It should always be there.

To be clear, any choice is fine, but I don't think the argument of better error message or better performance matter here, because it should never error.